### ISSUE TITLE
feat: Trace collector dev-tool for flagging successful/failed agent interactions

### ISSUE THREAD

**Issue by bitboody** [pewdiepie-archdaemon/odysseus#3060]
### Prerequisites

- [x] I searched [open issues](https://github.com/pewdiepie-archdaemon/odysseus/issues?q=is%3Aissue+is%3Aopen) and this has not already been proposed.
- [x] I searched [discussions](https://github.com/pewdiepie-archdaemon/odysseus/discussions) and this is not already being debated there.
- [x] This is a concrete, actionable proposal — not a vague "it would be nice if..." request.

### Area

Other

### Problem or Motivation

Per the [GitHub discussion](https://github.com/pewdiepie-archdaemon/odysseus/discussions/2724), this contributes to Felix’s proposed effort to collect training data for fine-tuning a model tailored specifically for Odysseus.

### Proposed Solution

Following the discussion around dataset collection for an Odysseus fine-tune, I think lowering the barrier to contribution and standardizing traces would make sample gathering much easier.

Plan:
* Add an **"Advanced Dev Tools"** setting that gives contributors, maintainers, and testers access to developer-focused utilities and debugging features.
* Add a way to mark one or multiple agent interactions as **successful** or **failed**, then export/save them as structured **JSON** or **CSV** traces for easier dataset collection and analysis.

### Are you willing to implement this?

Partially — I can help but need guidance

**Comment by undergroundrap** [comment]
Good direction. I think the key is to make the first slice a local, opt-in annotation/export tool rather than a broad automatic trace pipeline.

Implementation shape I’d suggest:

- Treat a trace as a reviewed turn/range, not the whole chat by default: user message id(s), assistant message id, assistant metadata (`tool_events`, metrics, model name/class), reviewer label (`success` / `failure`), tags, and optional notes.
- Anchor marks to persisted message ids + session id, not array positions, because sessions can be compacted, edited, deleted, or reordered.
- Keep every route owner-scoped. Mark/export should verify the current user owns the session/message, and incognito sessions should not be exportable.
- Prefer JSON/JSONL as the primary export. CSV is useful for spreadsheets, but it is lossy for nested tool events, metrics, sources, and multi-message turns.
- Do not export secrets by default. At minimum, include redaction metadata such as `redaction_version` / `redacted: true`, and strip obvious API keys, auth headers, tokens, private endpoint URLs, and local paths.
- Keep fine-tune dataset upload/collection out of the first PR. First PR can be backend-only: mark/unmark/list/export + focused tests. The Advanced Dev Tools UI can be a follow-up once the data model is stable.

Tests I’d want before calling it safe:

- cross-owner session/message mark/export is rejected;
- incognito traces are rejected or omitted;
- missing/deleted message ids return 404;
- JSONL export preserves tool event/metric structure;
- redaction removes obvious token/header/API-key shapes;
- success/failure labels and tags round-trip.

I’m happy to help with the backend route/test shape if useful. Main thing I’d avoid is manual log scraping or broad automatic collection; reviewed, opt-in trace marks should produce cleaner samples and avoid privacy surprises.

**Comment by bitboody** [comment]
I’d be happy to work on the scraping part — maybe we can split the work somehow. A maintainer could assign people to different parts/tasks.

**Comment by undergroundrap** [comment]
> I’d be happy to work on the scraping part — maybe we can split the work somehow. A maintainer could assign people to different parts/tasks.

Sounds good. I’m heading offline for the night, but I’m happy to split it up.

If you want to take the scraping/collection side, I can probably take the backend shape around trace marking/export, owner scoping, redaction, and tests. Or if you had a different split in mind, let me know what part you feel most confident owning and I can take the other half.

Before either of us opens a PR, it might be worth sketching the task split here so maintainers can steer us if they prefer a different order.

**Comment by bitboody** [comment]
> > I’d be happy to work on the scraping part — maybe we can split the work somehow. A maintainer could assign people to different parts/tasks.
> 
> Sounds good. I’m heading offline for the night, but I’m happy to split it up.
> 
> If you want to take the scraping/collection side, I can probably take the backend shape around trace marking/export, owner scoping, redaction, and tests. Or if you had a different split in mind, let me know what part you feel most confident owning and I can take the other half.
> 
> Before either of us opens a PR, it might be worth sketching the task split here so maintainers can steer us if they prefer a different order.

I will also be back in a few hours, but so far I have only written code to read the ```data.db``` file.

Later on, I will implement an export function to export the wanted data to a JSON file.

**Comment by bitboody** [comment]
Here's a part of the code I am working on:
```py
    for session in sessions:
        # Session properties
        session_properties = {
            "state": session['chat_state'],
            "name": session['name'],
            "id": session['id'],
            "model": session['model'],
            "message_count": session['message_count'],
            "created_at": session['created_at']
        }
        # Note: Chat state persistence is handled by the frontend.
        # This service fetches the current state (e.g., 'success' or 'failed') 
        # directly from the database for display.
```

I think the responsibility for writing the "state" of a chat should be delegated to the frontend, since deciding whether a chat is "successful" or "failed" will be done through the UI.

**Comment by undergroundrap** [comment]
@bitboody 

Nice, that makes sense as a starting point.

One design thing I’d be careful about: I’d avoid making this a direct `data.db` scraper if possible. Since Odysseus already has session/message ownership checks and SQLAlchemy models, I think the safer shape is an app-level backend route/service that reads `sessions` + `chat_messages` through the existing DB layer, verifies ownership, applies redaction, and then exports the trace.

Also, I agree the UI is where the user decides whether a turn/session was `success` or `failed`, but I’d treat the frontend as the input surface, not the source of truth. The frontend can send “mark this message/turn as success/failed,” but the backend should validate and persist that annotation. That way export works consistently and we can test it.

I don’t currently see `chat_state` as an existing `sessions` column on current `dev`, so if that is a new field we probably want either:

- a small dedicated trace/annotation table keyed by `owner`, `session_id`, and message ids; or
- an explicit migration/metadata path if maintainers prefer keeping it lighter.

My instinct is that a dedicated trace mark/annotation model is cleaner because it avoids mixing dataset-review state into normal chat session state.

If you want to keep working on the export shape, I can help with the backend route/test side: owner scoping, incognito rejection, missing/deleted message handling, and JSON/JSONL structure. That should let us split the work without stepping on each other.

**Comment by bitboody** [comment]
@undergroundrap 
Maybe a lighter approach here could be having the frontend provide the `success` / `failure` label (since that’s ultimately a user judgment), while the backend validates and persists it for export. That could avoid introducing additional tables in the first iteration if maintainers want to keep things lightweight.

**Comment by undergroundrap** [comment]
@bitboody 

Yeah, I think that’s a good lightweight v1.

I’m not attached to a new table for the first pass if maintainers prefer keeping the migration surface small. The key boundary I’d keep is:

- frontend provides the user judgment (`success` / `failed`, maybe tags/notes);
- backend validates the session/message ownership and allowed label values;
- backend persists the annotation;
- export derives the actual trace data from the DB, not from client-supplied payloads.

A minimal version could store the annotation on the persisted assistant message metadata, something like:

```json
{
  "trace": {
    "label": "success",
    "tags": ["tool-use", "good-final-answer"],
    "note": "...",
    "marked_by": "user",
    "marked_at": "..."
  }
}
```

Then export can find marked assistant messages and reconstruct the turn from the surrounding persisted messages plus existing assistant metadata like `tool_events`, model, token metrics, etc.

The guardrails I’d still keep for v1:

- validate label/tag shape server-side;
- verify the current user owns the session/message before writing;
- reject incognito/unpersisted messages;
- do not trust client-supplied model/tool/session data;
- apply redaction during export;
- add tests for cross-owner writes, missing message ids, invalid labels, and JSON/JSONL export.

That seems like a good compromise: lightweight storage now, but still safe and testable.

**Comment by bitboody** [comment]
@undergroundrap

This feels like a solid starting point.

I’ve already built a proof-of-concept for the backend exporter, and I’m working toward something aligned with the discussion.

A possible shape for the core function:

```py
def scrape_chat_data(
    session_id: str,
    message_ids: list[str],
    label: str,
    user_id: str
):
```

This would be called from a route handler, which then delegates to a lower-level export function:

```py
def export_json(file_name: str, data: list[dict]):
```

To make collaboration easier and avoid diverging implementations, I think it might be useful to start draft PRs early so we can align on architecture before fully finalizing the implementation.

What do you think?

**Comment by undergroundrap** [comment]
@bitboody 

Yeah, I think an early draft PR is a good idea. It gives everyone something concrete to review before the shape gets too locked in.

The split you described makes sense to me, with one small naming/ownership note: I’d avoid treating this as raw “scraping” long-term, because the important part is that it goes through the app’s normal ownership/redaction rules rather than reading arbitrary DB rows. Something like `build_trace_export(...)` or `collect_trace_records(...)` might describe the intent a little better.

For the route boundary, I’d also make sure `user_id` is derived server-side from the current request, not accepted from the frontend. The frontend can send `session_id`, `message_ids`, `label`, tags/notes, etc., but the backend should decide who the caller is and verify all selected messages belong to that user/session.

Maybe the shape becomes something like:

```py
def build_trace_export(
    *,
    session_id: str,
    message_ids: list[str],
    label: str,
    owner: str,
) -> list[dict]:
    ...
```

And for `export_json`, I’d be careful with `file_name`: probably sanitize it or have the route return a JSON/JSONL download response rather than writing to an arbitrary path.

Happy to review the draft PR once you open it. I can focus on the guardrail/test side: owner scoping, invalid labels, missing/deleted message ids, incognito/unpersisted messages, redaction, and making sure the exported records are derived from persisted DB data rather than client-supplied trace payloads.

**Comment by bitboody** [comment]
@undergroundrap 
 This should get us started: https://github.com/pewdiepie-archdaemon/odysseus/pull/3192
